File: settings/backups/client/lib/api/analytics.ts
/**
 * Analytics API Client (T192)
 * 
 * Provides methods for interacting with analytics and metrics endpoints.
 * Handles request/response parsing, error handling, and authentication.
 * 
 * Endpoints:
 * - POST /api/analytics/users/:userId/track - Track user activity
 * - GET /api/analytics/users/:userId/summary - Get user analytics summary
 * - GET /api/analytics/users/:userId/timeseries - Get user time series data
 * - POST /api/analytics/courses/:courseId/track - Track course activity
 * - GET /api/analytics/courses/:courseId/summary - Get course analytics summary
 * - GET /api/analytics/courses/:courseId/timeseries - Get course time series
 * - GET /api/analytics/courses/top - Get top performing courses
 * - POST /api/analytics/engagement/track - Track engagement activity
 * - GET /api/analytics/engagement/summary - Get engagement summary
 * - GET /api/analytics/engagement/timeseries - Get engagement time series
 * - GET /api/metrics - Get all metrics
 * - GET /api/metrics/:id - Get metric by ID
 * - GET /api/metrics/name/:name - Get metric by name
 * - GET /api/metrics/category/:category - Get metrics by category
 * - POST /api/metrics/refresh - Refresh all metrics
 */

import { getTokens } from '@/lib/storage/secure-storage';

// API Configuration
const API_BASE_URL = process.env.EXPO_PUBLIC_API_URL || 'http://localhost:8080';
const API_TIMEOUT = 10000; // 10 seconds

// Period types for analytics queries
export type AnalyticsPeriod = 'today' | 'yesterday' | '7days' | '30days' | '90days' | 'year';

// Metric format types
export type MetricFormat = 'number' | 'currency' | 'percentage' | 'duration';

// Metric trend direction
export type TrendDirection = 'up' | 'down' | 'neutral';

// Request Types
export interface TrackUserActivityRequest {
  login_count?: number;
  active_minutes?: number;
  courses_accessed?: number;
  quizzes_completed?: number;
  points_earned?: number;
  badges_earned?: number;
  posts_created?: number;
  comments_made?: number;
  likes_given?: number;
  shares_made?: number;
}

export interface TrackCourseActivityRequest {
  enrollments?: number;
  completions?: number;
  dropouts?: number;
  avg_completion_rate?: number;
  avg_rating?: number;
  total_revenue?: number;
}

export interface TrackEngagementActivityRequest {
  daily_active_users?: number;
  weekly_active_users?: number;
  monthly_active_users?: number;
  total_sessions?: number;
  avg_session_duration?: number;
  bounce_rate?: number;
  conversion_rate?: number;
}

export interface GetAnalyticsSummaryParams {
  period?: AnalyticsPeriod;
}

export interface GetAnalyticsTimeSeriesParams {
  period?: AnalyticsPeriod;
}

export interface GetTopCoursesParams {
  period?: AnalyticsPeriod;
  orderBy?: 'enrollments' | 'completions' | 'rating' | 'revenue';
  limit?: number;
}

export interface GetMetricsParams {
  limit?: number;
  offset?: number;
}

export interface GetMetricsByCategoryParams {
  limit?: number;
  offset?: number;
}

// Response Types
export interface UserAnalytics {
  id: string;
  tenant_id: string;
  user_id: string;
  date: string;
  login_count: number;
  active_minutes: number;
  courses_accessed: number;
  courses_completed: number;
  quizzes_attempted: number;
  quizzes_completed: number;
  points_earned: number;
  badges_earned: number;
  posts_created: number;
  comments_made: number;
  likes_given: number;
  shares_made: number;
  messages_sent: number;
  videos_watched: number;
  livestreams_joined: number;
  created_at: string;
  updated_at: string;
}

export interface UserAnalyticsSummary {
  total_logins: number;
  total_active_minutes: number;
  avg_active_minutes: number;
  total_courses_accessed: number;
  total_courses_completed: number;
  total_quizzes_attempted: number;
  total_quizzes_completed: number;
  quiz_completion_rate: number;
  total_points_earned: number;
  total_badges_earned: number;
  total_social_activity: number;
  period: string;
  start_date: string;
  end_date: string;
}

export interface CourseAnalytics {
  id: string;
  tenant_id: string;
  course_id: string;
  date: string;
  enrollments: number;
  completions: number;
  dropouts: number;
  avg_completion_rate: number;
  avg_rating: number;
  total_revenue: number;
  created_at: string;
  updated_at: string;
}

export interface CourseAnalyticsSummary {
  total_enrollments: number;
  total_completions: number;
  total_dropouts: number;
  avg_completion_rate: number;
  avg_rating: number;
  total_revenue: number;
  period: string;
  start_date: string;
  end_date: string;
}

export interface EngagementAnalytics {
  id: string;
  tenant_id: string;
  date: string;
  daily_active_users: number;
  weekly_active_users: number;
  monthly_active_users: number;
  total_sessions: number;
  avg_session_duration: number;
  bounce_rate: number;
  conversion_rate: number;
  created_at: string;
  updated_at: string;
}

export interface EngagementAnalyticsSummary {
  avg_daily_active_users: number;
  avg_weekly_active_users: number;
  avg_monthly_active_users: number;
  total_sessions: number;
  avg_session_duration: number;
  avg_bounce_rate: number;
  avg_conversion_rate: number;
  period: string;
  start_date: string;
  end_date: string;
}

export interface DashboardMetric {
  id: string;
  tenant_id: string;
  name: string;
  description?: string;
  value: number;
  previous_value?: number;
  change?: number;
  change_percentage?: number;
  trend?: TrendDirection;
  format?: MetricFormat;
  category?: string;
  updated_at: string;
}

export interface TopCourse {
  course_id: string;
  course_name?: string;
  total_enrollments: number;
  total_completions: number;
  avg_rating: number;
  total_revenue: number;
}

// API Client
class AnalyticsAPIClient {
  private baseUrl: string;
  private timeout: number;

  constructor(baseUrl: string = API_BASE_URL, timeout: number = API_TIMEOUT) {
    this.baseUrl = baseUrl;
    this.timeout = timeout;
  }

  /**
   * Make authenticated API request
   */
  private async request<T>(
    endpoint: string,
    options: RequestInit = {}
  ): Promise<T> {
    const tokens = await getTokens();
    
    const headers = {
      'Content-Type': 'application/json',
      ...(tokens?.accessToken && { Authorization: `Bearer ${tokens.accessToken}` }),
      ...options.headers,
    };

    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), this.timeout);

    try {
      const response = await fetch(`${this.baseUrl}${endpoint}`, {
        ...options,
        headers,
        signal: controller.signal,
      });

      clearTimeout(timeoutId);

      if (!response.ok) {
        const error = await response.json().catch(() => ({ error: 'Request failed' }));
        throw new Error(error.error || `HTTP ${response.status}`);
      }

      return await response.json();
    } catch (error) {
      if (error instanceof Error) {
        if (error.name === 'AbortError') {
          throw new Error('Request timeout');
        }
        throw error;
      }
      throw new Error('Unknown error occurred');
    }
  }

  // ==================== USER ANALYTICS ====================

  /**
   * Track user activity
   */
  async trackUserActivity(
    userId: string,
    data: TrackUserActivityRequest
  ): Promise<{ message: string }> {
    return this.request(`/api/analytics/users/${userId}/track`, {
      method: 'POST',
      body: JSON.stringify(data),
    });
  }

  /**
   * Get user analytics summary
   */
  async getUserAnalyticsSummary(
    userId: string,
    params?: GetAnalyticsSummaryParams
  ): Promise<{ summary: UserAnalyticsSummary }> {
    const query = new URLSearchParams();
    if (params?.period) query.set('period', params.period);

    return this.request(`/api/analytics/users/${userId}/summary?${query}`);
  }

  /**
   * Get user analytics time series
   */
  async getUserAnalyticsTimeSeries(
    userId: string,
    params?: GetAnalyticsTimeSeriesParams
  ): Promise<{ timeseries: UserAnalytics[] }> {
    const query = new URLSearchParams();
    if (params?.period) query.set('period', params.period);

    return this.request(`/api/analytics/users/${userId}/timeseries?${query}`);
  }

  // ==================== COURSE ANALYTICS ====================

  /**
   * Track course activity
   */
  async trackCourseActivity(
    courseId: string,
    data: TrackCourseActivityRequest
  ): Promise<{ message: string }> {
    return this.request(`/api/analytics/courses/${courseId}/track`, {
      method: 'POST',
      body: JSON.stringify(data),
    });
  }

  /**
   * Get course analytics summary
   */
  async getCourseAnalyticsSummary(
    courseId: string,
    params?: GetAnalyticsSummaryParams
  ): Promise<{ summary: CourseAnalyticsSummary }> {
    const query = new URLSearchParams();
    if (params?.period) query.set('period', params.period);

    return this.request(`/api/analytics/courses/${courseId}/summary?${query}`);
  }

  /**
   * Get course analytics time series
   */
  async getCourseAnalyticsTimeSeries(
    courseId: string,
    params?: GetAnalyticsTimeSeriesParams
  ): Promise<{ timeseries: CourseAnalytics[] }> {
    const query = new URLSearchParams();
    if (params?.period) query.set('period', params.period);

    return this.request(`/api/analytics/courses/${courseId}/timeseries?${query}`);
  }

  /**
   * Get top performing courses
   */
  async getTopCourses(
    params?: GetTopCoursesParams
  ): Promise<{ courses: TopCourse[] }> {
    const query = new URLSearchParams();
    if (params?.period) query.set('period', params.period);
    if (params?.orderBy) query.set('orderBy', params.orderBy);
    if (params?.limit) query.set('limit', params.limit.toString());

    return this.request(`/api/analytics/courses/top?${query}`);
  }

  // ==================== ENGAGEMENT ANALYTICS ====================

  /**
   * Track engagement activity
   */
  async trackEngagementActivity(
    data: TrackEngagementActivityRequest
  ): Promise<{ message: string }> {
    return this.request('/api/analytics/engagement/track', {
      method: 'POST',
      body: JSON.stringify(data),
    });
  }

  /**
   * Get engagement analytics summary
   */
  async getEngagementAnalyticsSummary(
    params?: GetAnalyticsSummaryParams
  ): Promise<{ summary: EngagementAnalyticsSummary }> {
    const query = new URLSearchParams();
    if (params?.period) query.set('period', params.period);

    return this.request(`/api/analytics/engagement/summary?${query}`);
  }

  /**
   * Get engagement analytics time series
   */
  async getEngagementAnalyticsTimeSeries(
    params?: GetAnalyticsTimeSeriesParams
  ): Promise<{ timeseries: EngagementAnalytics[] }> {
    const query = new URLSearchParams();
    if (params?.period) query.set('period', params.period);

    return this.request(`/api/analytics/engagement/timeseries?${query}`);
  }

  // ==================== METRICS ====================

  /**
   * Get all metrics
   */
  async getMetrics(
    params?: GetMetricsParams
  ): Promise<{ metrics: DashboardMetric[]; count: number }> {
    const query = new URLSearchParams();
    if (params?.limit) query.set('limit', params.limit.toString());
    if (params?.offset) query.set('offset', params.offset.toString());

    return this.request(`/api/metrics?${query}`);
  }

  /**
   * Get metric by ID
   */
  async getMetric(id: string): Promise<{ metric: DashboardMetric }> {
    return this.request(`/api/metrics/${id}`);
  }

  /**
   * Get metric by name
   */
  async getMetricByName(name: string): Promise<{ metric: DashboardMetric }> {
    return this.request(`/api/metrics/name/${name}`);
  }

  /**
   * Get metrics by category
   */
  async getMetricsByCategory(
    category: string,
    params?: GetMetricsByCategoryParams
  ): Promise<{ metrics: DashboardMetric[]; category: string; count: number }> {
    const query = new URLSearchParams();
    if (params?.limit) query.set('limit', params.limit.toString());
    if (params?.offset) query.set('offset', params.offset.toString());

    return this.request(`/api/metrics/category/${category}?${query}`);
  }

  /**
   * Refresh all metrics (admin only)
   */
  async refreshAllMetrics(): Promise<{ message: string }> {
    return this.request('/api/metrics/refresh', {
      method: 'POST',
    });
  }

  /**
   * Refresh metric by name (admin only)
   */
  async refreshMetricByName(name: string): Promise<{ metric: DashboardMetric; message: string }> {
    return this.request(`/api/metrics/name/${name}/refresh`, {
      method: 'POST',
    });
  }
}

// Export singleton instance
export const analyticsAPI = new AnalyticsAPIClient();

// Export types and client
export default analyticsAPI;

File: server/internal/models/analytics.go
package models

import (
	"time"

	"github.com/google/uuid"
	"gorm.io/gorm"
)

// UserAnalytics represents daily user activity metrics
type UserAnalytics struct {
	ID               uuid.UUID      `gorm:"type:uuid;primary_key" json:"id"`
	TenantID         uuid.UUID      `gorm:"type:uuid;not null;index:idx_user_analytics_tenant" json:"tenant_id"`
	UserID           uuid.UUID      `gorm:"type:uuid;not null;index:idx_user_analytics_user" json:"user_id"`
	Date             time.Time      `gorm:"type:date;not null;index:idx_user_analytics_date" json:"date"`
	LoginCount       int            `gorm:"default:0" json:"login_count"`
	ActiveMinutes    int            `gorm:"default:0" json:"active_minutes"`
	CoursesStarted   int            `gorm:"default:0" json:"courses_started"`
	CoursesCompleted int            `gorm:"default:0" json:"courses_completed"`
	LessonsCompleted int            `gorm:"default:0" json:"lessons_completed"`
	QuizzesAttempted int            `gorm:"default:0" json:"quizzes_attempted"`
	QuizzesPassed    int            `gorm:"default:0" json:"quizzes_passed"`
	PointsEarned     int            `gorm:"default:0" json:"points_earned"`
	BadgesEarned     int            `gorm:"default:0" json:"badges_earned"`
	PostsCreated     int            `gorm:"default:0" json:"posts_created"`
	CommentsCreated  int            `gorm:"default:0" json:"comments_created"`
	LikesGiven       int            `gorm:"default:0" json:"likes_given"`
	MessagesSet      int            `gorm:"default:0" json:"messages_sent"`
	CreatedAt        time.Time      `json:"created_at"`
	UpdatedAt        time.Time      `json:"updated_at"`
	DeletedAt        gorm.DeletedAt `gorm:"index" json:"deleted_at,omitempty"`
}

// BeforeCreate hook to set UUID
func (ua *UserAnalytics) BeforeCreate(tx *gorm.DB) error {
	if ua.ID == uuid.Nil {
		ua.ID = uuid.New()
	}
	return nil
}

// TableName specifies the table name
func (UserAnalytics) TableName() string {
	return "user_analytics"
}

// CourseAnalytics represents course performance metrics
type CourseAnalytics struct {
	ID               uuid.UUID      `gorm:"type:uuid;primary_key" json:"id"`
	TenantID         uuid.UUID      `gorm:"type:uuid;not null;index:idx_course_analytics_tenant" json:"tenant_id"`
	CourseID         uuid.UUID      `gorm:"type:uuid;not null;index:idx_course_analytics_course" json:"course_id"`
	Date             time.Time      `gorm:"type:date;not null;index:idx_course_analytics_date" json:"date"`
	TotalEnrollments int            `gorm:"default:0" json:"total_enrollments"`
	ActiveStudents   int            `gorm:"default:0" json:"active_students"`
	CompletionRate   float64        `gorm:"type:decimal(5,2);default:0" json:"completion_rate"`
	AverageProgress  float64        `gorm:"type:decimal(5,2);default:0" json:"average_progress"`
	AverageRating    float64        `gorm:"type:decimal(3,2);default:0" json:"average_rating"`
	TotalReviews     int            `gorm:"default:0" json:"total_reviews"`
	DropoutRate      float64        `gorm:"type:decimal(5,2);default:0" json:"dropout_rate"`
	AverageTimeSpent int            `gorm:"default:0" json:"average_time_spent"` // in minutes
	TotalRevenue     float64        `gorm:"type:decimal(10,2);default:0" json:"total_revenue"`
	CreatedAt        time.Time      `json:"created_at"`
	UpdatedAt        time.Time      `json:"updated_at"`
	DeletedAt        gorm.DeletedAt `gorm:"index" json:"deleted_at,omitempty"`
}

// BeforeCreate hook to set UUID
func (ca *CourseAnalytics) BeforeCreate(tx *gorm.DB) error {
	if ca.ID == uuid.Nil {
		ca.ID = uuid.New()
	}
	return nil
}

// TableName specifies the table name
func (CourseAnalytics) TableName() string {
	return "course_analytics"
}

// EngagementAnalytics represents platform-wide engagement metrics
type EngagementAnalytics struct {
	ID                     uuid.UUID      `gorm:"type:uuid;primary_key" json:"id"`
	TenantID               uuid.UUID      `gorm:"type:uuid;not null;index:idx_engagement_analytics_tenant" json:"tenant_id"`
	Date                   time.Time      `gorm:"type:date;not null;index:idx_engagement_analytics_date" json:"date"`
	TotalActiveUsers       int            `gorm:"default:0" json:"total_active_users"`
	NewUsers               int            `gorm:"default:0" json:"new_users"`
	ReturningUsers         int            `gorm:"default:0" json:"returning_users"`
	DailyActiveUsers       int            `gorm:"default:0" json:"daily_active_users"`
	WeeklyActiveUsers      int            `gorm:"default:0" json:"weekly_active_users"`
	MonthlyActiveUsers     int            `gorm:"default:0" json:"monthly_active_users"`
	AverageSessionDuration float64        `gorm:"type:decimal(10,2);default:0" json:"average_session_duration"` // in minutes
	TotalSessions          int            `gorm:"default:0" json:"total_sessions"`
	PageViews              int            `gorm:"default:0" json:"page_views"`
	BounceRate             float64        `gorm:"type:decimal(5,2);default:0" json:"bounce_rate"`
	ConversionRate         float64        `gorm:"type:decimal(5,2);default:0" json:"conversion_rate"`
	CreatedAt              time.Time      `json:"created_at"`
	UpdatedAt              time.Time      `json:"updated_at"`
	DeletedAt              gorm.DeletedAt `gorm:"index" json:"deleted_at,omitempty"`
}

// BeforeCreate hook to set UUID
func (ea *EngagementAnalytics) BeforeCreate(tx *gorm.DB) error {
	if ea.ID == uuid.Nil {
		ea.ID = uuid.New()
	}
	return nil
}

// TableName specifies the table name
func (EngagementAnalytics) TableName() string {
	return "engagement_analytics"
}

// DashboardMetric represents a single metric displayed on dashboards
type DashboardMetric struct {
	ID          uuid.UUID      `gorm:"type:uuid;primary_key" json:"id"`
	TenantID    uuid.UUID      `gorm:"type:uuid;not null;index:idx_dashboard_metric_tenant" json:"tenant_id"`
	Name        string         `gorm:"type:varchar(100);not null;index:idx_dashboard_metric_name" json:"name"`
	Value       float64        `gorm:"type:decimal(20,2);not null" json:"value"`
	Label       string         `gorm:"type:varchar(255);not null" json:"label"`
	Description string         `gorm:"type:text" json:"description,omitempty"`
	Category    string         `gorm:"type:varchar(50);index:idx_dashboard_metric_category" json:"category"`
	Format      string         `gorm:"type:varchar(50)" json:"format"` // number, percentage, currency, duration
	Unit        string         `gorm:"type:varchar(20)" json:"unit,omitempty"`
	Icon        string         `gorm:"type:varchar(50)" json:"icon,omitempty"`
	Trend       string         `gorm:"type:varchar(20)" json:"trend,omitempty"`    // up, down, stable
	Change      float64        `gorm:"type:decimal(10,2)" json:"change,omitempty"` // percentage change
	Target      float64        `gorm:"type:decimal(20,2)" json:"target,omitempty"`
	CreatedAt   time.Time      `json:"created_at"`
	UpdatedAt   time.Time      `json:"updated_at"`
	DeletedAt   gorm.DeletedAt `gorm:"index" json:"deleted_at,omitempty"`
}

// BeforeCreate hook to set UUID
func (dm *DashboardMetric) BeforeCreate(tx *gorm.DB) error {
	if dm.ID == uuid.Nil {
		dm.ID = uuid.New()
	}
	return nil
}

// TableName specifies the table name
func (DashboardMetric) TableName() string {
	return "dashboard_metrics"
}
